### DISCUSSION TITLE
Does the GS ALC app really need to be online all the time?

### DISCUSSION THREAD

**Discussion by Lineflyer** [cgeo/cgeo#10728] [Q&A] [answered]
Today I played an adventure lab (assisted by c:geo of course) and it was more or less the first time out of my homezone.
While playing I experienced some issues, which I did not really expect.

I could also look up some other forums but would like to get feedback here just out of curiosity and as I know @baiti has probably most experience.

- The main problem was, that although I selected to "Start" the adventure everything seems still online related. I realized this, because after stage 1 there was only 2G connection and the game was stuck. I could not even show the next stages information or click on answer. I would have expected, that this information is downloaded by the app when I start that adventure but not trying it on the fly?!
- The next problem seems related...after I managed to load the next stage and gave my answer, the app was unresponsive for nearly one minute until i got the confirmation of a correct answer. Before that I gave a wrong answer, this was immediately denied. What is this.....the answer seems to be local but on correct answer some things are loaded over the net?
- The two points above sum up to the following question:  Even if an online connection is needed, I still don't get the long response time.  I checked my internet speed, it was 700 kbit/s. Shouldn't this be more than enough to timely check answers and load some small graphics?

Summarizing....it looks nice and easy but its IMHO not well engineered. I never realized this before....but maybe you all know that?

**Comment by baiti** [comment, marked as answer]
As a geocacher, for a long time I stayed away from this game for various reasons most of which have been discussed in great detail in other forums. From a software development perspective I got interested in the game, some of the results you can see in the ALC integration in cgeo. Of course I also played some of the adventures using the regular player app and it appears this is indeed a game that requires a permanent online connection with reasonable quality in terms of bandwidth and throughput. I have read quite a few logs by now where geocachers complained about a terrible user experience due to limitied connectivity out in the field. So much for an intro, now to your questions specifically, based on the experience I have with that app.

When a green pin is touched in the app or the app is launched via the external link, the adventure is fetched from GC's servers. It comes in via a single API call as a JSON response, the size of such an object is a few KB at most, depending on how much descriptive text the owner provided at each stage. There is information in that object that provides the app with the capability to verify whether an answer given to a certain question is correct or not correct. No interaction with the server is required at that time. Wrong answers are rejected directly by the app. If a correct answer is given, the answer string is immediately sent to the server, the "found it" credit gets immediately awarded. This all is just a few KB in data transfer, nothing to worry about. Now it gets interesting. When this object is initially loaded, it contains links to futher information, most notably high resoution images, some of which are implicitly loaded by the app when it thinks it is appropriate, others (e.g. YouTube videos) are loaded at the user's explicit request it he chooses so.

* When the app is loaded, a `KeyImage` is fetched and displayed plus some descriptive text. It is also known as the `Cover Image`. Here is what is said about it in the builder: 

  > Choose an image that captures the theme of your Adventure. A simple, text-free image 1125x1050 pixels or larger is best as this image will be automatically cropped and used in several places on your Adventure.

* For each stage:
  * when you approach the stage, based on geofencing, a `KeyImage` needs to be loaded, more descriptive text and last not least, the question is renderered. The builder says again:
    > Choose an image to represent this location. Recommended size: 1125x1050 or larger.

  * Once you give a correct answer:
    * an `AwardImage` is fetched from the server (same assumptions about size as outlined above)
    * a completion message is shown (that is part of the object which was initially fetched when the adventure was loaded)

People usually use the images captured with their smartphones, most of the time these are hi res images in the order of 5MB to 10MB per image. I have not explicitly tested it but I am confident, the builder doesn't down scale these images when the adventure is created.  I am in the process of creating an adventure myself and I will for sure upload the highest quality images my SmartPhone offers.

This game won't work nicely in rural areas with limited connectivity, also better don't play if your mobile data is on a tight budget. Typically these adventures are played in cities where you have at least 4G these days. There are exceptions where those adventures take you out into the woods,  then the user experience is limited depending on where it actually is.

This game is for the always connected world where bandwith and volume of mobile data is not a concern.

**Reply by eddiemuc** [reply]
@baiti covered it ver nicely: yes ALCs are meant for the "online only" community.

@Lineflyer specifically you mentioned the long waiting time after typing in a correct answer vs the short time after an incorrect answer. I noticed this too, and if course the only possible explanation is that answers are checked on client/smartphone side (and the long wait after correct answer is probably the call to the server to mark the stage as found on server and increase your geocache found counter by one). But this also means that the app is designed really badly, because obviously they need to transfer the correct answer to the client before the user has answered. And this opens a wide door for fraud. (Unless they use hashes instead of the plain answer which I don't know)

**Reply by baiti** [reply]
> @baiti covered it ver nicely: yes ALCs are meant for the "online only" community.

;)

> @Lineflyer specifically you mentioned the long waiting time after typing in a correct answer vs the short time after an incorrect answer. I noticed this too, and if course the only possible explanation is that answers are checked on client/smartphone side (and 

Answer is checked on the phone that I am absolutely sure.

> the long wait after correct answer is probably the call to the server to mark the stage as found on server and increase your geocache found counter by one). 

No, that happens instantly, e.g. less than a second.

> But this also means that the app is designed really badly, because obviously they need to transfer the correct answer to the client before the user has answered. And this opens a wide door for fraud. (Unless they use hashes instead of the plain answer which I don't know)

You'd be surprised how smart they do this. For sure the correct answer is not just sent over the wire in clear text. It doesn't even help you to intercept the communication via man in the middle. They do use hashes, salted hashes to be precise.

The long delays come from loading images. Upon correct answering a stage, an "AwardImage" is loaded, this could be 10MB. Depends on the builder (owner), how fancy he/she wanted to be.

**Reply by Lineflyer** [reply]
Thanks for your replies.
Then it is indeed as I feared it would be. 

I know that this is an online game....however prefetching all media files as soon as I select "Start" for a specific adventure would IMHO be a better user experience ("Wait some seconds....we are loading that game..."). Because fetching/sending a few kB to verify the answer is less a problem, than loading the media files only at the point of time, where the correct answer is provided.

Also the user would have more trust, that he can actually play the game.
Yesterday I ended up with running around between an area with better coverage and the geofenced area to provide the answer.... :(


EDIT:
>  ("Wait some seconds....we are loading that game...")

Or even just do it in the background after selecting "Start".
But anyway....not our app. Just interesting to see, how easy things could be improved in that app.

**Reply by baiti** [reply]
> I know that this is an online game....however prefetching all media files as soon as I select "Start" for a specific adventure would IMHO be a better user experience ("Wait some seconds....we are loading that game..."). Because fetching/sending a few kB to verify the answer is less a problem, than loading the media files only at the point of time, where the correct answer is provided.

You are looking at from the same perspective as if you'd go on a regular caching tour. You'd prepare at home and have everything you need stored locally, most likely in cgeo ;)

This game is meant for ad hoc play. You stroll around in the city, no plan whatsoever, then you stumble over that ALC and then it is just all unravelling.

> 
> Also the user would have more trust, that he can actually play the game.
> Yesterday I ended up with running around between an area with better coverage and the geofenced area to provide the answer.... :(

I have read many complaints along those lines.

I have seen adventures where the owner set the geofence to 100km (! .. yes, this is officially supported) and set the stages to "random" so the player can actually and without any location spoofer play his own sequence and from everywhere he wants. Clever questions / answers make him go to the destination. This doesn't have to be enforced via some questionable geofencing.
